File: catty-web/packages/terminal-emulation/src/terminal/ColorResolver.ts
/**
 * Color Resolver
 * 
 * Utilities for resolving different color types to CSS values.
 * Handles ANSI colors, 256-color palette, and RGB colors.
 */

import type { SgrColorType, SgrNamedColor } from './TerminalEmulationTypes';

/**
 * Standard 16 ANSI color names mapped to CSS variable names
 */
export const ANSI_COLOR_VARIABLES: Record<SgrNamedColor, string> = {
  black: 'var(--terminal-color-black)',
  red: 'var(--terminal-color-red)',
  green: 'var(--terminal-color-green)',
  yellow: 'var(--terminal-color-yellow)',
  blue: 'var(--terminal-color-blue)',
  magenta: 'var(--terminal-color-magenta)',
  cyan: 'var(--terminal-color-cyan)',
  white: 'var(--terminal-color-white)',
  brightBlack: 'var(--terminal-color-bright-black)',
  brightRed: 'var(--terminal-color-bright-red)',
  brightGreen: 'var(--terminal-color-bright-green)',
  brightYellow: 'var(--terminal-color-bright-yellow)',
  brightBlue: 'var(--terminal-color-bright-blue)',
  brightMagenta: 'var(--terminal-color-bright-magenta)',
  brightCyan: 'var(--terminal-color-bright-cyan)',
  brightWhite: 'var(--terminal-color-bright-white)',
};

/**
 * 256-color palette generator
 * Generates the standard xterm 256-color palette
 */
export class Color256Palette {
  private static readonly STANDARD_COLORS = [
    // Standard 16 colors (0-15) - use CSS variables
    'var(--terminal-color-black)',
    'var(--terminal-color-red)',
    'var(--terminal-color-green)',
    'var(--terminal-color-yellow)',
    'var(--terminal-color-blue)',
    'var(--terminal-color-magenta)',
    'var(--terminal-color-cyan)',
    'var(--terminal-color-white)',
    'var(--terminal-color-bright-black)',
    'var(--terminal-color-bright-red)',
    'var(--terminal-color-bright-green)',
    'var(--terminal-color-bright-yellow)',
    'var(--terminal-color-bright-blue)',
    'var(--terminal-color-bright-magenta)',
    'var(--terminal-color-bright-cyan)',
    'var(--terminal-color-bright-white)',
  ];

  /**
   * Get color value for 256-color palette index
   * @param index Color index (0-255)
   * @returns CSS color value
   */
  public static getColor(index: number): string {
    // Standard 16 colors (0-15)
    if (index >= 0 && index <= 15) {
      return this.STANDARD_COLORS[index];
    }

    // 216 color cube (16-231)
    if (index >= 16 && index <= 231) {
      return this.getCubeColor(index - 16);
    }

    // Grayscale ramp (232-255)
    if (index >= 232 && index <= 255) {
      return this.getGrayscaleColor(index - 232);
    }

    // Invalid index, return default
    return 'var(--terminal-foreground)';
  }

  /**
   * Generate color from 6x6x6 color cube
   * @param cubeIndex Index in the color cube (0-215)
   * @returns RGB color string
   */
  private static getCubeColor(cubeIndex: number): string {
    const r = Math.floor(cubeIndex / 36);
    const g = Math.floor((cubeIndex % 36) / 6);
    const b = cubeIndex % 6;

    const toColorValue = (n: number) => n === 0 ? 0 : 55 + n * 40;

    return `rgb(${toColorValue(r)}, ${toColorValue(g)}, ${toColorValue(b)})`;
  }

  /**
   * Generate grayscale color
   * @param grayIndex Index in grayscale ramp (0-23)
   * @returns RGB color string
   */
  private static getGrayscaleColor(grayIndex: number): string {
    const gray = 8 + grayIndex * 10;
    return `rgb(${gray}, ${gray}, ${gray})`;
  }
}

/**
 * Color resolver for different SGR color types
 */
export class ColorResolver {
  /**
   * Resolve SGR color type to CSS color value
   * @param colorType SGR color specification
   * @returns CSS color value
   */
  public static resolve(colorType: SgrColorType): string {
    switch (colorType.type) {
      case 'named':
        return this.resolveNamedColor(colorType.color);
      case 'indexed':
        return this.resolveIndexedColor(colorType.index);
      case 'rgb':
        return this.resolveRgbColor(colorType.r, colorType.g, colorType.b);
      default:
        return 'var(--terminal-foreground)';
    }
  }

  /**
   * Resolve named ANSI color to CSS variable
   */
  private static resolveNamedColor(color: SgrNamedColor): string {
    return ANSI_COLOR_VARIABLES[color] || 'var(--terminal-foreground)';
  }

  /**
   * Resolve indexed color (256-color palette)
   */
  private static resolveIndexedColor(index: number): string {
    return Color256Palette.getColor(index);
  }

  /**
   * Resolve RGB color to CSS rgb() value
   */
  private static resolveRgbColor(r: number, g: number, b: number): string {
    // Clamp values to valid range
    const clampedR = Math.max(0, Math.min(255, r));
    const clampedG = Math.max(0, Math.min(255, g));
    const clampedB = Math.max(0, Math.min(255, b));
    
    return `rgb(${clampedR}, ${clampedG}, ${clampedB})`;
  }

  /**
   * Convert ANSI color code to SgrColorType
   * @param colorCode ANSI color code (30-37, 40-47, 90-97, 100-107)
   * @param isForeground Whether this is a foreground color
   * @returns SgrColorType or null if invalid
   */
  public static ansiCodeToColorType(colorCode: number): SgrColorType | null {
    // Standard foreground colors (30-37)
    if (colorCode >= 30 && colorCode <= 37) {
      const colors: SgrNamedColor[] = [
        'black', 'red', 'green', 'yellow',
        'blue', 'magenta', 'cyan', 'white'
      ];
      return { type: 'named', color: colors[colorCode - 30] };
    }

    // Bright foreground colors (90-97)
    if (colorCode >= 90 && colorCode <= 97) {
      const colors: SgrNamedColor[] = [
        'brightBlack', 'brightRed', 'brightGreen', 'brightYellow',
        'brightBlue', 'brightMagenta', 'brightCyan', 'brightWhite'
      ];
      return { type: 'named', color: colors[colorCode - 90] };
    }

    // Standard background colors (40-47)
    if (colorCode >= 40 && colorCode <= 47) {
      const colors: SgrNamedColor[] = [
        'black', 'red', 'green', 'yellow',
        'blue', 'magenta', 'cyan', 'white'
      ];
      return { type: 'named', color: colors[colorCode - 40] };
    }

    // Bright background colors (100-107)
    if (colorCode >= 100 && colorCode <= 107) {
      const colors: SgrNamedColor[] = [
        'brightBlack', 'brightRed', 'brightGreen', 'brightYellow',
        'brightBlue', 'brightMagenta', 'brightCyan', 'brightWhite'
      ];
      return { type: 'named', color: colors[colorCode - 100] };
    }

    return null;
  }

  /**
   * Create indexed color type
   * @param index Color index (0-255)
   * @returns SgrColorType for indexed color
   */
  public static createIndexedColor(index: number): SgrColorType {
    return { type: 'indexed', index };
  }

  /**
   * Create RGB color type
   * @param r Red component (0-255)
   * @param g Green component (0-255)
   * @param b Blue component (0-255)
   * @returns SgrColorType for RGB color
   */
  public static createRgbColor(r: number, g: number, b: number): SgrColorType {
    return { type: 'rgb', r, g, b };
  }

  /**
   * Create named color type
   * @param color Named color
   * @returns SgrColorType for named color
   */
  public static createNamedColor(color: SgrNamedColor): SgrColorType {
    return { type: 'named', color };
  }
}

File: catty-ksa/caTTY.Display/Rendering/ColorResolver.cs
using caTTY.Core.Types;
using Brutal.Numerics;

namespace caTTY.Display.Rendering;

/// <summary>
/// Color resolver for different SGR color types.
/// Handles ANSI colors, 256-color palette, and RGB colors.
/// Based on TypeScript ColorResolver implementation.
/// </summary>
public static class ColorResolver
{
    // Note: Hardcoded colors removed - now using ThemeManager for dynamic theme support

    /// <summary>
    /// Resolve SGR color type to ImGui float4 color value.
    /// </summary>
    /// <param name="color">SGR color specification</param>
    /// <param name="isBackground">Whether this is a background color (affects defaults)</param>
    /// <returns>ImGui float4 color value</returns>
    public static float4 Resolve(caTTY.Core.Types.Color? color, bool isBackground = false)
    {
        if (!color.HasValue)
        {
            return isBackground ? ThemeManager.GetDefaultBackground() : ThemeManager.GetDefaultForeground();
        }

        return color.Value.Type switch
        {
            ColorType.Named => ResolveNamedColor(color.Value.NamedColor),
            ColorType.Indexed => ResolveIndexedColor(color.Value.Index),
            ColorType.Rgb => ResolveRgbColor(color.Value.Red, color.Value.Green, color.Value.Blue),
            _ => isBackground ? ThemeManager.GetDefaultBackground() : ThemeManager.GetDefaultForeground()
        };
    }

    /// <summary>
    /// Resolve named ANSI color to float4 color using current theme.
    /// </summary>
    private static float4 ResolveNamedColor(NamedColor namedColor)
    {
        return namedColor switch
        {
            NamedColor.Black => ThemeManager.ResolveThemeColor(0),
            NamedColor.Red => ThemeManager.ResolveThemeColor(1),
            NamedColor.Green => ThemeManager.ResolveThemeColor(2),
            NamedColor.Yellow => ThemeManager.ResolveThemeColor(3),
            NamedColor.Blue => ThemeManager.ResolveThemeColor(4),
            NamedColor.Magenta => ThemeManager.ResolveThemeColor(5),
            NamedColor.Cyan => ThemeManager.ResolveThemeColor(6),
            NamedColor.White => ThemeManager.ResolveThemeColor(7),
            NamedColor.BrightBlack => ThemeManager.ResolveThemeColor(8),
            NamedColor.BrightRed => ThemeManager.ResolveThemeColor(9),
            NamedColor.BrightGreen => ThemeManager.ResolveThemeColor(10),
            NamedColor.BrightYellow => ThemeManager.ResolveThemeColor(11),
            NamedColor.BrightBlue => ThemeManager.ResolveThemeColor(12),
            NamedColor.BrightMagenta => ThemeManager.ResolveThemeColor(13),
            NamedColor.BrightCyan => ThemeManager.ResolveThemeColor(14),
            NamedColor.BrightWhite => ThemeManager.ResolveThemeColor(15),
            _ => ThemeManager.GetDefaultForeground()
        };
    }

    /// <summary>
    /// Resolve indexed color (256-color palette) to float4 color using current theme for ANSI colors.
    /// </summary>
    private static float4 ResolveIndexedColor(byte index)
    {
        // Standard 16 colors (0-15) - use theme colors
        if (index <= 15)
        {
            return ThemeManager.ResolveThemeColor(index);
        }

        // 216 color cube (16-231)
        if (index >= 16 && index <= 231)
        {
            return GetCubeColor(index - 16);
        }

        // Grayscale ramp (232-255)
        if (index >= 232 && index <= 255)
        {
            return GetGrayscaleColor(index - 232);
        }

        // Invalid index, return default
        return ThemeManager.GetDefaultForeground();
    }

    /// <summary>
    /// Generate color from 6x6x6 color cube.
    /// </summary>
    /// <param name="cubeIndex">Index in the color cube (0-215)</param>
    /// <returns>float4 color</returns>
    private static float4 GetCubeColor(int cubeIndex)
    {
        int r = cubeIndex / 36;
        int g = (cubeIndex % 36) / 6;
        int b = cubeIndex % 6;

        static float ToColorValue(int n) => n == 0 ? 0.0f : (55 + n * 40) / 255.0f;

        return new float4(ToColorValue(r), ToColorValue(g), ToColorValue(b), 1.0f);
    }

    /// <summary>
    /// Generate grayscale color.
    /// </summary>
    /// <param name="grayIndex">Index in grayscale ramp (0-23)</param>
    /// <returns>float4 color</returns>
    private static float4 GetGrayscaleColor(int grayIndex)
    {
        float gray = (8 + grayIndex * 10) / 255.0f;
        return new float4(gray, gray, gray, 1.0f);
    }

    /// <summary>
    /// Resolve RGB color to float4 color.
    /// </summary>
    private static float4 ResolveRgbColor(byte red, byte green, byte blue)
    {
        return new float4(red / 255.0f, green / 255.0f, blue / 255.0f, 1.0f);
    }

    /// <summary>
    /// Convert ANSI color code to Color type.
    /// </summary>
    /// <param name="colorCode">ANSI color code (30-37, 40-47, 90-97, 100-107)</param>
    /// <returns>Color or null if invalid</returns>
    public static caTTY.Core.Types.Color? AnsiCodeToColor(int colorCode)
    {
        // Standard foreground colors (30-37)
        if (colorCode >= 30 && colorCode <= 37)
        {
            var namedColors = new[]
            {
                NamedColor.Black, NamedColor.Red, NamedColor.Green, NamedColor.Yellow,
                NamedColor.Blue, NamedColor.Magenta, NamedColor.Cyan, NamedColor.White
            };
            return new caTTY.Core.Types.Color(namedColors[colorCode - 30]);
        }

        // Bright foreground colors (90-97)
        if (colorCode >= 90 && colorCode <= 97)
        {
            var namedColors = new[]
            {
                NamedColor.BrightBlack, NamedColor.BrightRed, NamedColor.BrightGreen, NamedColor.BrightYellow,
                NamedColor.BrightBlue, NamedColor.BrightMagenta, NamedColor.BrightCyan, NamedColor.BrightWhite
            };
            return new caTTY.Core.Types.Color(namedColors[colorCode - 90]);
        }

        // Standard background colors (40-47)
        if (colorCode >= 40 && colorCode <= 47)
        {
            var namedColors = new[]
            {
                NamedColor.Black, NamedColor.Red, NamedColor.Green, NamedColor.Yellow,
                NamedColor.Blue, NamedColor.Magenta, NamedColor.Cyan, NamedColor.White
            };
            return new caTTY.Core.Types.Color(namedColors[colorCode - 40]);
        }

        // Bright background colors (100-107)
        if (colorCode >= 100 && colorCode <= 107)
        {
            var namedColors = new[]
            {
                NamedColor.BrightBlack, NamedColor.BrightRed, NamedColor.BrightGreen, NamedColor.BrightYellow,
                NamedColor.BrightBlue, NamedColor.BrightMagenta, NamedColor.BrightCyan, NamedColor.BrightWhite
            };
            return new caTTY.Core.Types.Color(namedColors[colorCode - 100]);
        }

        return null;
    }

    /// <summary>
    /// Create indexed color type.
    /// </summary>
    /// <param name="index">Color index (0-255)</param>
    /// <returns>Color for indexed color</returns>
    public static caTTY.Core.Types.Color CreateIndexedColor(byte index)
    {
        return new caTTY.Core.Types.Color(index);
    }

    /// <summary>
    /// Create RGB color type.
    /// </summary>
    /// <param name="r">Red component (0-255)</param>
    /// <param name="g">Green component (0-255)</param>
    /// <param name="b">Blue component (0-255)</param>
    /// <returns>Color for RGB color</returns>
    public static caTTY.Core.Types.Color CreateRgbColor(byte r, byte g, byte b)
    {
        return new caTTY.Core.Types.Color(r, g, b);
    }

    /// <summary>
    /// Create named color type.
    /// </summary>
    /// <param name="namedColor">Named color</param>
    /// <returns>Color for named color</returns>
    public static caTTY.Core.Types.Color CreateNamedColor(NamedColor namedColor)
    {
        return new caTTY.Core.Types.Color(namedColor);
    }
}